File: easy/1137-tribonacci.py
# 1137. N-th Tribonacci Number

# strat: DP
# see C++ solution for even faster one

class Solution(object):
    def tribonacci(self, n):
        memo = {}

        def countStairs(m): 
            if m == 1:
                return 1
            if m == 0:
                return 0
            if m == 2:
                return 1
            if m in memo: 
                return memo[m]
            needed = countStairs(m-1) + countStairs(m-2) + countStairs(m-3)
            memo[m] = needed
            return needed
        
        return countStairs(n)

File: easy/1137-tribonacci.cpp
// 1137. N-th Tribonacci Number

// fast solution
// strat: use closed form for the linear recurrence

#include <cmath>
class Solution {
    double a = (std::cbrt(19 + 3 * std::sqrt(33)) + std::cbrt(19 - 3 * std::sqrt(33)) + 1)/3;
    double b = std::cbrt(586 + 102 * std::sqrt(33));
    double t = 3*b/(b*b - 2*b + 4);
public:
    int tribonacci(int n) {
        return std::round(t*std::pow(a,n));
    }
};